src/core/Akka.Streams/Stage/GraphStage.cs

```diff
@@ -423,6 +425,16 @@ public Timer(int id, ICancelable task)
     /// </summary>
     public abstract class GraphStageLogic : IStageLogging
     {
+        private List<Func<object, Task>>? CallbacksWaitingForInterpreter = null;
+        private AtomicReference<List<TaskCompletionSource<Done>>?> AsyncCallbacksInProgress = new(new List<TaskCompletionSource<Done>>());
```

Comment:
This holds the current async callback invocations that are currently in progress. Note that it is a nullable atomic reference to a list of `TaskCompletionSource`s. We use null to signal that the stream is shutting down or in the process of doing so, all callback invocations will automatically trigger a `StreamDetachedException` if it sees that this list is null. We need this because we need to guarantee cross-thread access safety (async callbacks are invoked from outside the stream)